specs/GetFavicon.md

```diff
@@ -0,0 +1,155 @@
+/// The 'error_code` is E_NOT_SET if the there is no image. Otherwise error_code 
+/// is the result from the image write operation.
+/// For more details, see the `GetFavicon` API.
+[uuid(A2508329-7DA8-49D7-8C05-FA125E4AEE8D), object, pointer_default(unique)]
+interface ICoreWebView2GetFaviconCompletedHandler : IUnknown {
+  /// Called to notify the favicon has been retrieved.
+  HRESULT Invoke([in] HRESULT error_code);
+}
+
+/// This is the ICoreWebView2 Experimental Favicon interface.
+[uuid(DC838C64-F64B-4DC7-98EC-0992108E2157), object, pointer_default(unique)]
+interface ICoreWebView2_10 : ICoreWebView2_9 {
+    /// Add an event handler for the `FaviconChanged` event.
+    /// The `FaviconChanged` event is raised when the 
+    /// [favicon](https://developer.mozilla.org/en-US/docs/Glossary/Favicon)
+    /// of the top-level document changes or if script dynamically changes the favicon.
+    /// The FaviconChanged event will be raised for first navigating to a new 
+    /// document, whether or not a document declares a Favicon in HTML. The event will 
+    /// be raised again if a favicon is declared in its HTML or has script 
+    /// to set its favicon. The favicon information can then be retrieved with 
+    /// `GetFavicon` and `FaviconUri`.
+    HRESULT add_FaviconChanged(
+        [in] ICoreWebViewFaviconChangedEventHandler* eventHandler,
+        [out] EventRegistrationToken* token);
+
+    /// Remove the event handler for `FaviconChanged` event.
+    HRESULT remove_FaviconChanged(
+        [in] EventRegistrationToken token);
+
+    /// Get the current Uri of the favicon as a string.
+    /// If the value is null, then the return value is `E_POINTER`, otherwise it is `S_OK`.
+    /// If a page has no favicon then the value is an empty string.
+    [propget] HRESULT FaviconUri([out, retval] LPWSTR* value);
+
+    /// Async function for getting the actual image data of the favicon.
+    /// If the `imageStream` is null, the `HRESULT` will be `E_POINTER`, otherwise
+    /// it is `S_OK`.
+    /// The image is copied to the `imageStream` object. If there is no image then
+    /// no data would be copied into the imageStream.
+    /// In either scenario the `completedHandler` is executed at the end of the operation.
+    HRESULT GetFavicon(
+        [in] COREWEBVIEW2_FAVICON_IMAGE_FORMAT format,
+        [in] IStream* imageStream,
+        [in] ICoreWebView2GetFaviconCompletedHandler* completedHandler);
+}
+
+[v1_enum]
+typedef enum COREWEBVIEW2_FAVICON_IMAGE_FORMAT {
+    /// Indicates that CoreWebView2.GetFaviconAsync should return the favicon in PNG format.
+    COREWEBVIEW2_CAPTURE_PREVIEW_IMAGE_FORMAT_PNG,
+
+    /// Indicates that CoreWebView2.GetFaviconAsync should return the favicon in JPG format.
+    COREWEBVIEW2_CAPTURE_PREVIEW_IMAGE_FORMAT_JPEG,
+}
+```
+
+## .Net/ WinRT
+```c#
+namespace Microsoft.Web.WebView2.Core
+{
+    enum CoreWebView2FaviconImageFormat
```

Comment:
https://en.wikipedia.org/wiki/Favicon#File_format_support says Favicons are supported in ICO, SVG, and GIF.  Do we plan to extend this enum if this API becomes popular?

Reply:
I believe that (@mrsteamfist correct me) WebView2 will support favicon images in the same format as Edge supports which includes all the image formats you mention. This API lets you get the image back out of WebView2 only in these two formats. That is, WebView2 can read an SVG favicon from the web, stores an internal raw representation of that image internally, and then later the CoreWebView2.GetFaviconAsync API can get you that favicon converting from its internal raw representation to png or jpeg.

As mentioned above we need to update the docs since we've gotten two questions about this.

Reply:
@MikeHillberg But that creates problems when we add support for a new image format. Suppose we add SVG support, and somebody calls GetFavicon. Should we return (stream, SVG)? But the caller might not support SVG (since SVG didn't exist at the time they were written).

Reply:
Agreed. Also, then might have to come up with a name for the return tuple :). I don't remember other image APIs that have this built-in conversion, so I was surprised to see this API so advanced. Also, the current shape of the API doesn't allow a caller to get out an image format that it understands and WV2 doesn't.

Reply:
> Would it be easier to just return the image data and format and let the app deal with it?

Its not easier for our implementation. We're building an API on existing internals which watch for changes to the favicon and convert to a raw internal representation. I don't know that we can easily obtain the original bytes or format. But we do also have the property to provide the URI of the favicon if the end dev wants to do something wildly different from what most would want via GetFavicon.

And I also would imagine its less convenient for the caller since they have to handle more image formats.

Reply:
The data for the FavIcon is translated into a unique data structure. The format has to be translated again to consumption by the user. This is a required step since these images are stored in a database.